Bug Fix && Content block updates

mendelB opened the pull request:
## _PULL REQUEST OVERVIEW_


### What does this PR do?
This PR
- [900d122](https://github.com/DoSomething/phoenix-next/commit/900d1222d87d41c60a218ab4af0f293acd12a7ce) adds a `campaignUpdate` case to our Contentful Entry block parser, since that type is allowed as an `affirmationBlock` field in `ShareActions`
- [f89b4b](https://github.com/DoSomething/phoenix-next/commit/f89b4b22f4ab41280f2921a16f44846388f19688) updates our `ContentModal` to allow for general `block` content, doing the proper entry lookup and using the `ContentfulEntry` component to properly render the `json` fields.
- [d50b170](https://github.com/DoSomething/phoenix-next/commit/d50b1705e58b103c347e93837659f4a2004cce76) updates the `CampaignUpdate` component to not override it's props in the container, by manually performing the entry lookup.


### Any background context you want to provide?
I bumped into this issue trying to test out the post share action affirmation modal, and attaching a `campaignUpdate` as the `affirmationBlock`. This KO'd the app since we didn't parse the block as a CampaignUpdate. Upon fixing this, I bumped into some other issues with campaignUpdates props being overridden in the `CampaignUpdate` component which led to some necessary updates to the `ContentModal` so that we no longer need to do those overrides, and to bring it up to speed with our new 'general block' `ContentfulEntry` abilities.

weerd commented on resources/assets/components/CampaignUpdate/CampaignUpdateContainer.js:
```diff
@@ -1,24 +1,17 @@
-    affiliateLogo,
-    author,
-    content,
-    link,
```
Nice cleanup!

mendelB commented on resources/assets/components/Modal/containers/ContentModalContainer.js:
```diff
@@ -4,34 +4,13 @@ import ContentModal from '../configurations/ContentModal';
-  ];
+  const json = find(state.campaign.pages, { id })
+    || find(state.campaign.actionSteps, { id })
+    || find(state.campaign.activityFeed, { id });
```
We seem to be repeating this entry lookup a bunch. I imagine it might be nice to consolidate them all in a helper function?

DFurnes replied:
I'm into that. 👍

DFurnes commented on resources/assets/components/Modal/configurations/ContentModal.js:
```diff
@@ -1,48 +1,25 @@
+import ContentfulEntry from '../../ContentfulEntry';
+import ModalControls from '../ModalControls';
 
 const ContentModal = (props) => {
```
Do we want to rename this now that it doesn't _just_ hold a CampaignUpdate? `BlockModal`?

mendelB replied:
I dig.

weerd commented on resources/assets/components/ContentfulEntry/ContentfulEntry.js:
```diff
@@ -42,10 +42,12 @@ const ContentfulEntry = ({ json = DEFAULT_BLOCK, stepIndex, isSignedUp }: Conten
           content={json.fields.content}
           displayOptions={json.fields.displayOptions}
           link={json.fields.link}
+          bordered={false}
```
The no bordered will just be for when it appears in a modal right? But in the page feed it's still bordered?

mendelB replied:
Hmm, I believe this is the only place that we're rendering a CampaignUpdate from now so it'd always not be bordered

mendelB replied:
I'll try and fix that real quick

weerd replied:
Ok, yeah, cause they should be bordered in the feed... 😬

DFurnes replied:
Could we unify the styles so they look okay in both places?

Maybe a like a `rgba(0, 0, 0, 0.15)` border instead of `#dcdcdc`?

DFurnes replied:
ALSO, LOOK AT THOSE LITTLE GITHUB COLOR PREVIEWS. WOW. 🎨

mendelB replied:
whoah. Yeah unifying would be dope, since it'd be challenging to account for whether it's a modal vs feed cuz unless we want to tamper with the json or start looking at the window location?

weerd replied:
yeah, was thinking maybe that could be an option too so they don't need to have subtle differences in style.

also, the GitHub color previews are so cool! haha!

mendelB replied:
It seems to still look the same, unfortunately:
![image](https://user-images.githubusercontent.com/12417657/37118767-4ef94dc2-2223-11e8-9e08-d79d16d109a6.png)

mendelB replied:
Perhaps it's a cop-out or just ingenious you decide. But changing the border to a darker color from within the `modal.scss` file -- only for modal `article`s does the trick!

weerd replied:
Oh gotcha. Yeah, it's because it's a transparent value but the background it's transparent on is the card (which is white) and not the modal.

DFurnes replied:
Try `background-clip: padding-box` on the card!

mendelB replied:
By Jove! That works!

mendelB commented:
UPDATES:
- renamed the `ContentModal` to `BlockModal` and adjusted the calls to said modal type all over the place
- added a specific `border-color` rule for `<article>`s in modals to address the border issue. (and removed the prop in `ContentfulEntry` so it'll work as it did before in the feed etc.)
- added a helper method to find contentful entries by id from state. (replacing the three places we were calling this method with the helper call)
- using the `BlockModal` from within the `PostSignupModal` for a quick win 🎐

mendelB commented:
![image](https://user-images.githubusercontent.com/12417657/37121370-dcfb0a5a-222a-11e8-9cd5-94fa9d24a61a.png)

weerd commented on app/Entities/Entity.php:
```diff
@@ -68,6 +68,8 @@ public function parseBlock($block)
             case 'affirmation':
                 return new Affirmation($block->entry);
+            case 'campaignUpdate':
+                return new CampaignUpdate($block->entry);
```
This was killing me on Teens For Jeans test campaign, haha!

DFurnes commented on resources/assets/components/Modal/configurations/BlockModal.js:
```diff
@@ -0,0 +1,27 @@
+const BlockModal = (props) => {
+  const { json, closeModal } = props;
+
+  json.fields.bordered = false;
```
Can we add a `// HACK:` comment here explaining this?

weerd replied:
oh yeah, forgot about this.  @mendelB I think lets go with @DFurnes suggestion of using rgba value and not forcing removing the border. I'd rather get into a pattern of finding design solution that appease the blocks showing up in Modals and on the Page without changes like this if possible :)

mendelB replied:
Ugh that was left in there from when I was tinkering around 🤒

mendelB replied:
https://github.com/DoSomething/phoenix-next/pull/737#discussion_r172987979

weerd commented on resources/assets/helpers/index.js:
```diff
@@ -438,6 +438,20 @@ export function query(key, url = window.location) {
+ * @param  {String} id
+ * @return {Object|Boolean}
+ */
+export function findContentfulEntry(state, id) {
```
Noice!

DFurnes commented on resources/assets/components/Modal/modal.scss:
```diff
@@ -31,6 +31,10 @@
     }
+
+    article {
+      border-color: $black;
```
What's this for? Can we add a comment, or a more descriptive class to hook onto?

mendelB replied:
It's the solution to this: https://github.com/DoSomething/phoenix-next/pull/737#discussion_r172987979

I'll add in a clarifying comment!

DFurnes replied:
Oh I see that now! Let's definitely add a `// HACK:` describing that & we can revisit sometime later.

DFurnes commented on resources/assets/helpers/index.js:
```diff
@@ -438,6 +438,20 @@ export function query(key, url = window.location) {
 
+/**
+ * Find an entry from within the campaign by given ID param.
+ * (Returns false if not found).
```
Is this accurate? It should return `undefined` or `null`, ideally.

mendelB replied:
I think `find` itself returns `undefined` but since we have the `||` operators it renders as a boolean, or truthy value if found.

Should we change this to be found value or `undefined` if not found?

DFurnes replied:
Can you double-check that with a `console.log` or testing on the console? If all of the pieces of this boolean expression return `undefined`, it _shooooould_ return the last value as far as I know.

mendelB replied:
You're absolutely right, it does return `undefined`!
Here's what tripped me up:
![image](https://user-images.githubusercontent.com/12417657/37169127-05e6d692-22d5-11e8-8c18-22e6e36ddde4.png)

DFurnes replied:
I think the console is tricking you into using this [window.find](https://developer.mozilla.org/en-US/docs/Web/API/Window/find) method instead!

mendelB replied:
Aha! That makes sense, and is something to bear in mind for the future!

mendelB commented:
UPDATES:
- return value spec in docblock for `findContentfulEntry`
- use `background-clip: border-box` and dave's magic rgba value for `.bordered` instead of a modal hack
- remove a thing

weerd commented:
> use background-clip: border-box

Oh yeah! Nice 💃
